src/compiler/program.ts

```diff
@@ -904,7 +904,13 @@ namespace ts {
             let declarationDiagnostics: Diagnostic[] = [];
 
             if (options.noEmit) {
-                return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emittedFiles: undefined, emitSkipped: true };
+                // Allow only emitting of declarations
+                if (options.declaration) {

```

Comment:
I think it makes sense for declaration:true to override noEmit. Since it's something you have to do explicitly. I.e just having noEmit will emit nothing. If you really really want just declarations which is a valid use case then declarations:true and noEmit:true should give you the result.